app/jobs/autoscaling.py

```diff
@@ -156,13 +188,15 @@ def update_metric_values(asg, value, my_environment):
     return results
 
 
-def count_healthy_instances(asg):
-    # This will become more useful when we start scaling down
+def count_healthy_instances(asg, tag_list, draining_tag):
+    draining_instance_ids = get_draining_servers(asg, tag_list, draining_tag).keys()
+    num_draining = len(draining_instance_ids)
     num_healthy = 0
     for inst in asg["Instances"]:
-        if inst["HealthStatus"] == "Healthy" and inst["LifecycleState"] != "Terminating":
+        if inst["InstanceId"] not in draining_instance_ids and \
+            inst["ProtectedFromScaleIn"] and inst["HealthStatus"] == "Healthy" and inst["LifecycleState"] != "Terminating":
             num_healthy += 1
-    return num_healthy
+    return num_healthy, num_draining
```

Comment:
It's hard to track the invariants and progress properties of this scaling operation when there are so many flags and tags being checked.   One of the advantages of modeling the life of an instance with a state machine is that you can then use the state machine state definitions to describe the state of the distributed system -- see there I used "state" with two different meanings in the same sentence, so evil.   The point is, you can say there are 4 states of an instance, call them S1, S2, S3, S4 and you can count the instances in each state using a single counting function.   Then any decision you make ought to be based on s1_count, s2_count, s3_count, and s4_count.  I believe when you structure the code in this way, the reasoning for its correctness will also become simpler.  I defer to you if/when you would consider such restructuring.